fix: create one kubeconfig context per destination

mxyng opened the pull request:
## Summary

<!-- Include a summary of the change and/or why it's necessary. -->

create only one kubeconfig context per destination regardless if there may be multiple namespaces granted to the user.

update `use` to set the `Namespace` field of the target context.

Resolves #3611

mxyng commented on internal/cmd/kubernetes.go:
```diff
@@ -85,80 +85,78 @@ func writeKubeconfig(user *api.User, destinations []api.Destination, grants []ap
 
-		u, err := urlx.Parse(url)
+		contextName := "infra:" + cluster
+		if _, ok := infraContexts[contextName]; ok && namespace != "" {
```
If there's a non-namespace grant, set that as the default context. Otherwise, pick a namespace

BruceMacD reviewed (commented):
Looks good once checks pass

dnephin reviewed (commented):
I have a hard time following `writeKubeconfig`.

Do we have a test case for each of these cases that you could point me at?
1. a user with access to multiple namespaces gets a single context with no namespace set
2. a user with access to a single namespace gets a single context with the namespace set
3. a user with access to the full cluster (not a single namespace) gets a single context with no namespace set

mxyng commented:
Good call out on the tests. It was definitely missing some cases. I've refactored the test so it should be easier to follow and added the missing cases

dnephin commented on internal/cmd/kubernetes_test.go:
```diff
@@ -48,123 +43,123 @@ func TestWriteKubeconfig(t *testing.T) {
+			"infra:connected": {
+				AuthInfo:  "user",
+				Cluster:   "infra:connected",
+				Namespace: "namespace",
```
I thought the expectation would be for `Namespace: ""`, since this user has access to multiple namespaces. Is this test expecting the right thing?

mxyng replied:
`Namespace: ""` is equivalent to `Namespace: "default"` which the user may not have access to. It makes sense to set `Namespace` to something as long as the user doesn't have some form of global access. We can probably introduce heuristics to make the selection smarter but for now, it'll pick the first namespace user has access to.

dnephin replied:
I guess if one of the namespaces they have access to is default, then it would make sense for that to be the namespace. If they don't have access to default then any arbitrary one seems reasonable.

dnephin reviewed (commented):
Nice! The tests look great now! I'll try and give `writeKubeconfig` a more detailed read tomorrow. One question about the tests

jmorganca commented on internal/cmd/kubernetes.go:
```diff
@@ -85,80 +85,82 @@ func writeKubeconfig(user *api.User, destinations []api.Destination, grants []ap
+			continue
+		}
+
+		contextName := "infra:" + cluster
```
nit: could check this and `continue` earlier in the function

jmorganca commented on internal/cmd/kubernetes.go:
```diff
@@ -85,80 +85,82 @@ func writeKubeconfig(user *api.User, destinations []api.Destination, grants []ap
 	}
 
-	keep := make(map[string]bool)
+	type clusterContext struct {
```
```suggestion
	type context struct {
```

mxyng replied:
I wanted to avoid clobbering the `context` package

jmorganca replied:
Good call!

jmorganca commented on internal/cmd/kubernetes.go:
```diff
@@ -85,80 +85,82 @@ func writeKubeconfig(user *api.User, destinations []api.Destination, grants []ap
+		CA        []byte
+	}
+
+	infraContexts := make(map[string]clusterContext)
```
```suggestion
	contexts := make(map[string]context)
```

jmorganca commented on internal/cmd/kubernetes.go:
```diff
@@ -85,80 +85,82 @@ func writeKubeconfig(user *api.User, destinations []api.Destination, grants []ap
+		infraContexts[contextName] = infraContext
+	}
+
+	for contextName, infraContext := range infraContexts {
```
```suggestion
	for name, context := range contexts {
```

jmorganca reviewed (commented):
Nit: some of the naming could be simplified/tightened

jmorganca commented:
Looks good! Some super minor comments
